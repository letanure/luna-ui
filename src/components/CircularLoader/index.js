import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import hex2Rgba from '../../ultils/hex2Rgba';

/**
 * CircularLoader component.
 * E.g.:
 * ```html
 * <CircularLoader type="emit" kind="default" size="100px" />
 * ```
 */
class CircularLoader extends React.PureComponent {
  getLoaderAnimation(props) {
    return `
      @keyframes rotate-animation {
        0% {
          transform: rotate(0deg);
        }
        100% {
          transform: rotate(360deg);
        }
      }

      @keyframes anti-rotate-animation {
        0% {
          transform: rotate(0deg);
        }
        100% {
          transform: rotate(-360deg);
        }
      }

      :before {
        width: ${this.props.size - (this.props.border * 2)}px;
        height: ${this.props.size - (this.props.border * 2)}px;
        border-bottom-color: ${props.theme.circularLoader[this.props.kind]};
        border-right-color: ${props.theme.circularLoader[this.props.kind]};
        border-top-color: ${hex2Rgba(`${props.theme.circularLoader[this.props.kind]}`, 0)};
        border-left-color: ${hex2Rgba(`${props.theme.circularLoader[this.props.kind]}`, 0)};
        top: 0px;
        left: 0px;
        animation: rotate-animation 1s linear 0s infinite;
      }

      :after {
        width: ${(this.props.size - (this.props.border * 2)) * 0.7}px;
        height: ${(this.props.size - (this.props.border * 2)) * 0.7}px;
        border-bottom-color: ${props.theme.circularLoader[this.props.kind]};
        border-right-color: ${props.theme.circularLoader[this.props.kind]};
        border-top-color: ${hex2Rgba(`${props.theme.circularLoader[this.props.kind]}`, 0)};
        border-left-color: ${hex2Rgba(`${props.theme.circularLoader[this.props.kind]}`, 0)};
        top:  ${
  ((this.props.size - (this.props.border * 2)) - ((this.props.size - (this.props.border * 2)) * 0.7)) / 2
}px;
        left: ${
  ((this.props.size - (this.props.border * 2)) - ((this.props.size - (this.props.border * 2)) * 0.7)) / 2
}px;
        animation: anti-rotate-animation 0.85s linear 0s infinite;
      }
    `;
  }

  render() {
    const Loader = styled.div`
      position: relative;
      width: ${this.props.size}px;
      height: ${this.props.size}px;

      :before, :after {
        content: "";
        display: block;
        position: absolute;
        border-width: 4px;
        border-style: solid;
        border-radius: 50%;
      }

      ${(props) => this.getLoaderAnimation(props)}
    `;

    return <Loader />;
  }
}

CircularLoader.propTypes = {
  size: PropTypes.number,
  border: PropTypes.number,
  kind: PropTypes.string
};

CircularLoader.defaultProps = {
  size: 125,
  border: 4,
  kind: 'black'
};

export default CircularLoader;
