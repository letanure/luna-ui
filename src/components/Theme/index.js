import React from 'react';
import PropTypes from 'prop-types';
import { ThemeProvider } from 'styled-components';

import defaultTheme from '../../config/defaultTheme';
import DefaultStyles from '../../config/defaultStyles';
import DefaulTyphography from '../../config/defaultTyphography';

/**
 * Theme component.
 * E.g.:
 * ```html
 * <Theme theme={{...}}>
 *  ...
 * </Theme>
 * ```
 */
class ThemeComponent extends React.PureComponent {
  render() {
    const { theme, children } = this.props;

    return (
      <ThemeProvider theme={theme}>
        <div>
          <DefaultStyles />
          <DefaulTyphography />
          {children}
        </div>
      </ThemeProvider>
    );
  }
}

ThemeComponent.propTypes = {
  theme: PropTypes.object,
  children: PropTypes.node
};

ThemeComponent.defaultProps = {
  theme: defaultTheme,
  children: null
};

export default ThemeComponent;
