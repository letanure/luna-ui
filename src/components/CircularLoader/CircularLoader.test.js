import React from 'react';
import renderer from 'react-test-renderer';

import Theme from '../../config/defaultTheme';
import ThemeProvider from '../Theme/index';

import CircularLoader from './index';

describe('<CircularLoader />', () => {
  const component =
    <ThemeProvider theme={Theme}>
      <CircularLoader type="emit" kind="default" size={100} />
    </ThemeProvider>;

  test('toMatchSnapshot', () => {
    const tree = renderer
      .create(component)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
