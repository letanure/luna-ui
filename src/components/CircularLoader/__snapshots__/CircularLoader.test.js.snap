// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CircularLoader /> toMatchSnapshot 1`] = `
<div>
  <div
    className="sc-bdVaJa bAxxMn"
  />
</div>
`;
